File: onion-architecture/java/src/test/java/com/example/domain/services/EnrolmentServiceTest.java
package com.example.domain.services;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.example.domain.Course;
import com.example.domain.Student;

class EnrolmentServiceTest {

    private EnrolmentService enrolmentService;
    private Student student;
    private Course course;

    @BeforeEach
    void setUp() {
        enrolmentService = new EnrolmentService();
        student = new Student(1L, "John Doe");
        course = new Course(1L, "Java Programming");
    }

    @Test
    void shouldEnrolStudentInCourse() {
        enrolmentService.enrolStudent(student, course);

        assertTrue(student.getEnrolledCourses().contains(course));
        assertEquals(1, student.getEnrolledCourses().size());
    }

    @Test
    void shouldEnrolStudentInMultipleCourses() {
        Course secondCourse = new Course(2L, "Spring Boot");

        enrolmentService.enrolStudent(student, course);
        enrolmentService.enrolStudent(student, secondCourse);

        assertTrue(student.getEnrolledCourses().contains(course));
        assertTrue(student.getEnrolledCourses().contains(secondCourse));
        assertEquals(2, student.getEnrolledCourses().size());
    }

    @Test
    void shouldThrowExceptionWhenEnrollingStudentInSameCourseAgain() {
        enrolmentService.enrolStudent(student, course);

        IllegalStateException exception = assertThrows(
                IllegalStateException.class,
                () -> enrolmentService.enrolStudent(student, course)
        );

        assertEquals("Already enrolled in course", exception.getMessage());
        assertEquals(1, student.getEnrolledCourses().size());
    }

    @Test
    void shouldHandleMultipleStudentsEnrollingInSameCourse() {
        Student secondStudent = new Student(2L, "Jane Doe");

        enrolmentService.enrolStudent(student, course);
        enrolmentService.enrolStudent(secondStudent, course);

        assertTrue(student.getEnrolledCourses().contains(course));
        assertTrue(secondStudent.getEnrolledCourses().contains(course));
    }

    @Test
    void shouldNotAffectOtherStudentsWhenEnrolmentFails() {
        Student secondStudent = new Student(2L, "Jane Doe");

        enrolmentService.enrolStudent(student, course);
        enrolmentService.enrolStudent(secondStudent, course);

        assertThrows(
                IllegalStateException.class,
                () -> enrolmentService.enrolStudent(student, course)
        );

        assertTrue(secondStudent.getEnrolledCourses().contains(course));
        assertEquals(1, secondStudent.getEnrolledCourses().size());
    }
}

File: onion-architecture/kotlin/src/test/kotlin/com/example/domain/services/EnrolmentServiceTest.kt
package com.example.domain.services

import com.example.domain.Course
import com.example.domain.Student
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Assertions.*

class EnrolmentServiceTest {

    private lateinit var enrolmentService: EnrolmentService
    private lateinit var student: Student
    private lateinit var course: Course

    @BeforeEach
    fun setUp() {
        enrolmentService = EnrolmentService()
        student = Student(1L, "John Doe")
        course = Course(1L, "Java Programming")
    }

    @Test
    fun shouldEnrolStudentInCourse() {
        enrolmentService.enrolStudent(student, course)
        
        assertTrue(student.getEnrolledCourses().contains(course))
        assertEquals(1, student.getEnrolledCourses().size)
    }

    @Test
    fun shouldEnrolStudentInMultipleCourses() {
        val secondCourse = Course(2L, "Spring Boot")
        
        enrolmentService.enrolStudent(student, course)
        enrolmentService.enrolStudent(student, secondCourse)
        
        assertTrue(student.getEnrolledCourses().contains(course))
        assertTrue(student.getEnrolledCourses().contains(secondCourse))
        assertEquals(2, student.getEnrolledCourses().size)
    }

    @Test
    fun shouldThrowExceptionWhenEnrollingStudentInSameCourseAgain() {
        enrolmentService.enrolStudent(student, course)
        
        val exception = assertThrows(IllegalStateException::class.java) {
            enrolmentService.enrolStudent(student, course)
        }
        
        assertEquals("Already enrolled in course", exception.message)
        assertEquals(1, student.getEnrolledCourses().size)
    }

    @Test
    fun shouldHandleMultipleStudentsEnrollingInSameCourse() {
        val secondStudent = Student(2L, "Jane Doe")
        
        enrolmentService.enrolStudent(student, course)
        enrolmentService.enrolStudent(secondStudent, course)
        
        assertTrue(student.getEnrolledCourses().contains(course))
        assertTrue(secondStudent.getEnrolledCourses().contains(course))
    }

    @Test
    fun shouldNotAffectOtherStudentsWhenEnrolmentFails() {
        val secondStudent = Student(2L, "Jane Doe")
        
        enrolmentService.enrolStudent(student, course)
        enrolmentService.enrolStudent(secondStudent, course)
        
        assertThrows(IllegalStateException::class.java) {
            enrolmentService.enrolStudent(student, course)
        }
        
        assertTrue(secondStudent.getEnrolledCourses().contains(course))
        assertEquals(1, secondStudent.getEnrolledCourses().size)
    }
}